File: frontend/src/utils/orderItems.ts
// Groups order items by parent_order_item_id (add-ons nest under their
// container/parent) for a clearer receipt-style listing — mirrors
// bot/src/order_items.py so the bot and the app profile show the same
// structure. Uses the existing formatPriceK ("15к") compact price notation.
import { formatPriceK } from '@/utils/price'

export interface OrderItemLike {
    order_item_id: number
    product_id: number
    product_name?: string | null
    unit_price: number
    quantity: number
    parent_order_item_id?: number | null
}

export type OrderItemRow =
    | { type: 'flat'; key: number; text: string }
    | { type: 'group'; key: number; header: string; children: { key: number; text: string }[] }

const cleanName = (name: string) => name.replace(/\r\n|\r|\n/g, ' ').trim()

export function groupOrderItems(
    items: OrderItemLike[],
    productFallback: (productId: number) => string
): OrderItemRow[] {
    const childrenByParent = new Map<number, OrderItemLike[]>()
    for (const item of items) {
        if (item.parent_order_item_id != null) {
            const list = childrenByParent.get(item.parent_order_item_id) ?? []
            list.push(item)
            childrenByParent.set(item.parent_order_item_id, list)
        }
    }

    const rows: OrderItemRow[] = []
    for (const item of items) {
        if (item.parent_order_item_id != null) continue // rendered under its parent below

        const name = cleanName(item.product_name || productFallback(item.product_id))
        const kids = childrenByParent.get(item.order_item_id)

        if (kids && kids.length > 0) {
            const ownTotal = item.unit_price * item.quantity
            const header = ownTotal === 0 ? name : `${item.quantity} × ${name} - ${formatPriceK(ownTotal)}`
            rows.push({
                type: 'group',
                key: item.order_item_id,
                header,
                children: kids.map((child) => {
                    const childName = cleanName(child.product_name || productFallback(child.product_id))
                    return {
                        key: child.order_item_id,
                        text: `${child.quantity} × ${childName} - ${formatPriceK(child.unit_price * child.quantity)}`,
                    }
                }),
            })
        } else {
            rows.push({
                type: 'flat',
                key: item.order_item_id,
                text: `${item.quantity} × ${name} - ${formatPriceK(item.unit_price * item.quantity)}`,
            })
        }
    }
    return rows
}

File: bot/src/order_items.py
"""Shared item-list formatting for bot messages (admin cards, receipt caption,
client order-created message) — grouped by parent/add-on, compact "15к" price
notation. Mirrors frontend/src/utils/price.ts's formatPriceK for consistency
between the bot and the Mini App profile.
"""

import html
import re

_WHITESPACE_RUN = re.compile(r"\s+")


def format_price_k(amount: float | int) -> str:
    n = int(round(amount))
    if abs(n) >= 1_000_000:
        return f"{n:,}".replace(",", " ")
    return f"{round(n / 1000)}к"


def _clean_name(name: str) -> str:
    # Some product names carry an embedded newline (DB data quirk, e.g.
    # "Кастрюля 3.6\r\nJeep") — collapse ANY run of whitespace (space, \r, \n,
    # \r\n together) to a single space for compact list rendering, then escape
    # (parse_mode=HTML — this is client-controlled-ish data via the product
    # catalog, escape defensively regardless).
    collapsed = _WHITESPACE_RUN.sub(" ", str(name)).strip()
    return html.escape(collapsed)


def format_order_items(items: list[dict]) -> str:
    """Group items by parent_order_item_id: a container/parent with add-ons
    renders as a header line (name, plus its own qty/price if it costs
    something) followed by its children indented underneath; a standalone item
    (no add-ons) is just a flat bullet line. Returns the joined text (no
    trailing newline).
    """
    children: dict[int, list[dict]] = {}
    for item in items:
        parent_id = item.get("parent_order_item_id")
        if parent_id is not None:
            children.setdefault(parent_id, []).append(item)

    lines: list[str] = []
    for item in items:
        if item.get("parent_order_item_id") is not None:
            continue  # rendered under its parent below
        name = _clean_name(item.get("product_name") or f"Товар #{item['product_id']}")
        kids = children.get(item.get("order_item_id"))
        if kids:
            own_total = item["unit_price"] * item["quantity"]
            header = name if own_total == 0 else f"{name} ×{item['quantity']} - {format_price_k(own_total)}"
            lines.append(header)
            for child in kids:
                child_name = _clean_name(child.get("product_name") or f"Товар #{child['product_id']}")
                child_total = child["unit_price"] * child["quantity"]
                lines.append(f"   • {child_name} ×{child['quantity']} - {format_price_k(child_total)}")
        else:
            total = item["unit_price"] * item["quantity"]
            lines.append(f"  • {name} ×{item['quantity']} - {format_price_k(total)}")
    return "\n".join(lines)
